### PR TITLE
Support otel trace mapping

### PR DESCRIPTION
**Pull request by YANG-DB**
### Description
Simple Schema for Observability - SSO - will support the 3 major Observability types:
 - Traces
 - Logs
 - Metrics

SSO will comply with the [OTEL](https://github.com/open-telemetry) schema and continue supporting proprietary index structure

### Issues Resolved
- https://github.com/opensearch-project/observability/issues/1388

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: schema/traces/README.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+* Parent's Span identifier.
+* Links to causally-related Spans (via the SpanContext of those related Spans).
+* SpanContext information required to reference a Span. 
+
+### SpanContext
+Represents all the information that identifies Span in the Trace and is propagated to child Spans and across process boundaries.
+A **SpanContext** contains the tracing identifiers and the options that are propagated from parent to child Spans.
+
+* TraceId  - It is worldwide unique with practically sufficient probability by being made as 16 randomly generated bytes - used to group all spans for a specific trace together across all processes. 
+* SpanId  -  It is the identifier for a span, globally unique with practically sufficient probability by being made as 8 randomly generated bytes. When passed to a child Span this identifier becomes the parent span id for the child Span.
```
Comment (Swiddis) [inline schema/traces/README.md:71]: Do we need different randomness entropies? Does a simple `UUID` not suffice for both?

File: schema/traces/README.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+- **DroppedAttributesCount** : Integer counting the dropped attributes
+- **Events**     : A set of the next tuples (timestamp, name, Attributes)
+- **DroppedEventsCount** : Integer counting the dropped events
+- **Links**  : links to causally-related Spans
+- **DroppedLinksCount** : Integer counting the dropped links
+- **Status** - 
+  - UNSET= 0  -> The default status.
+  - OK= 1     -> The operation has been validated by an Application developer or Operator to have completed successfully.
+  - ERROR= 2  -> The operation contains an error.
```
Comment (Swiddis) [inline schema/traces/README.md:104]: [The spec](https://github.com/open-telemetry/opentelemetry-specification/blob/b00980832b4b823155001df56dbf9203d4e53f98/specification/trace/api.md#set-status) says there should be a precedence of `Ok > Error > Unset`. The current assignment makes implementing this logic needlessly tricky (should swap ok/error), unless these values are controlled by someone else.

Comment (derek-ho) [reply schema/traces/README.md:104]: I'm not sure precedence means that OK enum value has to be greater than ERROR enum value, it just means that if ok status is set >  more precedence than  > error status set  > more precedence than unset

File: schema/traces/README.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+- **EndTime**   : End time of the event
+- **Attributes**
+  - An Attribute is a key-value pair, which has the following structure  [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/b00980832b4b823155001df56dbf9203d4e53f98/specification/common/README.md#attribute)
+
+- **DroppedAttributesCount** : Integer counting the dropped attributes
+- **Events**     : A set of the next tuples (timestamp, name, Attributes)
+- **DroppedEventsCount** : Integer counting the dropped events
+- **Links**  : links to causally-related Spans
+- **DroppedLinksCount** : Integer counting the dropped links
+- **Status** -
```
Comment (derek-ho) [inline schema/traces/README.md:101]: I don't know if you need to include it here, but maybe separate out status.code and status.message?

File: schema/traces/traces.mapping [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+      "properties": {
+        "traceId": {
+          "ignore_above": 256,
+          "type": "keyword"
+        },
+        "spanId": {
+          "ignore_above": 256,
+          "type": "keyword"
+        },
+        "traceState": {
```
Comment (MaxKsyunz) [inline schema/traces/traces.mapping:64]: This doesn't look right. Should `traceState` be object or nested?

README.md states `traceState` is a key-value pairs, `traces.schema` lists them as object.

File: schema/traces/README.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+This is a logical unit of the application with which the emitted telemetry can be associated. It is typically the developer’s choice to decide what denotes a reasonable instrumentation scope.
+The most common approach is to use the instrumentation library as the scope, however other scopes are also common, e.g. a module, a package, or a class can be chosen as the instrumentation scope.
+
+The instrumentation scope may have zero or more additional attributes that provide additional information about the scope. As an example the field
+`instrumentationScope.attributes.identification` is presented will be used to determine the resource origin of the signal and can be used to filter accordingly
+
+## Traces
+see [OTEL traces convention](https://github.com/open-telemetry/opentelemetry-specification/tree/main/semantic_conventions/trace)
+
+Traces in are defined implicitly by their Spans - In particular, a Trace can be thought of as a directed acyclic graph (DAG) of Spans, where the edges between Spans are defined as parent/child relationship.
```
Comment (MaxKsyunz) [inline schema/traces/README.md:53]:
```suggestion
Traces in ???? are defined implicitly by their Spans - In particular, a Trace can be thought of as a directed acyclic graph (DAG) of Spans, where the edges between Spans are defined as parent/child relationship.
```

Word missing here.

File: schema/traces/traces.mapping [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+                  "ignore_above": 256
+                }
+              }
+            }
+          }
+        },
+        "instrumentationScope": {
+          "properties": {
+            "name": {
+              "type": "text",
```
Comment (MaxKsyunz) [inline schema/traces/traces.mapping:173]: keyword here?

File: schema/traces/traces.mapping [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+              "fields": {
+                "keyword": {
+                  "type": "keyword",
+                  "ignore_above": 128
+                }
+              }
+            },
+            "version": {
+              "type": "text",
```
Comment (MaxKsyunz) [inline schema/traces/traces.mapping:182]: keyword here as well?

File: schema/traces/traces.mapping [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+                "keyword": {
+                  "type": "keyword",
+                  "ignore_above": 256
+                }
+              }
+            }
+          }
+        },
+        "schemaUrl": {
+          "type": "text",
```
Comment (MaxKsyunz) [inline schema/traces/traces.mapping:205]: keyword?

File: schema/traces/samples/traceA.json [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  "droppedAttributesCount": 0,
+  "droppedEventsCount": 0,
+  "droppedLinksCount": 0,
+  "resource": {
+    "telemetry@sdk@name": "opentelemetry",
+    "telemetry@sdk@language": "python",
+    "telemetry@sdk@version": "0.14b0",
+    "service@name": "frontend-client",
+    "host@hostname": "ip-172-31-10-8.us-west-2.compute.internal"
+  },
```
Comment (MaxKsyunz) [inline schema/traces/samples/traceA.json:19]: How are these attribute keys generated? I do not see reference to dedoting in the readme or implied in the mapping or schema.

File: schema/traces/traces.mapping [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+          "properties": {
+            "name": {
+              "ignore_above": 1024,
+              "type": "keyword"
+            },
+            "@timestamp": {
+              "type": "date_nanos"
+            },
+            "observedTimestamp": {
+              "type": "date_nanos"
```
Comment (MaxKsyunz) [inline schema/traces/traces.mapping:144]: @YANG-DB  could you clarify difference between `@timestamp` and `observedTimestamp`?

I understand it as "`@timestamp` is value provided by opentelemetry" vs "`observedTimestamp` is the time exporter read the event record"

File: schema/traces/README.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+Additional fields can be supported via the Attributes key/value store see [traces](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/README.md)
+
+### Structure
+The default fields that are supported by the traces are
+- **TraceId**  : It is worldwide unique with practically sufficient probability by being made as 16 randomly generated bytes - used to group all spans for a specific trace together across all processes.
+- **SpanId**   : It is the identifier for a span, globally unique with practically sufficient probability by being made as 8 randomly generated bytes. When passed to a child Span this identifier becomes the parent span id for the child Span.
+- **ParentId** : It is the identifier for a span's parent span.
+- **TraceState** : carries tracing-system specific context in a list of key value pairs. Tracestate allows different vendors propagate additional information and inter-operate with their legacy Id formats. For more details see this.
```
Comment (MaxKsyunz) [inline schema/traces/README.md:81]: Missing link in 'For more details see this'?

Do you intend traceState to be sent unpacked? Otel just passes it around as a string.

File: schema/traces/README.md (hunk 1 above)
Comment (YANG-DB) [reply schema/traces/README.md:71]: this is the OTEL spec - we just publish the spec as is...

File: schema/traces/README.md (hunk 2 above)
Comment (YANG-DB) [reply schema/traces/README.md:104]: this is the spec in the OTEL trace protobuff:
https://github.com/open-telemetry/opentelemetry-proto/blob/e6dfd7f4cefdc2e8844fa3dfcd8b307477d59937/opentelemetry/proto/trace/v1/trace.proto#L264

File: schema/traces/README.md (hunk 3 above)
Comment (YANG-DB) [reply schema/traces/README.md:101]: will add the needed description...

File: schema/traces/traces.mapping (hunk 4 above)
Comment (YANG-DB) [reply schema/traces/traces.mapping:64]: i'll update the schema doc

File: schema/traces/samples/traceA.json (hunk 9 above)
Comment (YANG-DB) [reply schema/traces/samples/traceA.json:19]: this is generated from data-prepper, we will currently not assume anything about the transformation ...

File: schema/traces/traces.mapping (hunk 8 above)
Comment (YANG-DB) [reply schema/traces/traces.mapping:205]: yes

File: schema/traces/traces.mapping (hunk 7 above)
Comment (YANG-DB) [reply schema/traces/traces.mapping:182]: yes

File: schema/traces/traces.mapping (hunk 6 above)
Comment (YANG-DB) [reply schema/traces/traces.mapping:173]: thanks 🙏

File: schema/traces/traces.mapping (hunk 10 above)
Comment (YANG-DB) [reply schema/traces/traces.mapping:144]: yes - I'll add this to the README...

File: schema/traces/README.md (hunk 2 above)
Comment (YANG-DB) [reply schema/traces/README.md]: I've also added the status / message to the readme for further clarification

File: schema/traces/README.md (hunk 5 above)
Comment (YANG-DB) [reply schema/traces/README.md]: done - thanks

**Review by anirudha — APPROVED** [review]

**Review by ps48 — APPROVED** [review]
The BWC tests are failing as link to 2.6.0 artifact is not working. This may be due to build failures in jenkins ci [caused by other plugins or upstream]

**Comment by YANG-DB** [comment]
Irrelevant since introducing of Integration as part of 2.x branch

**Pull request closed without merging**